### PR TITLE
Async Chunk Loading and Generation

### PR DESCRIPTION
**Pull request by aikar**
# Async Chunk Loading and Generation - BETA

**STAGE:**
STABLE: Should be safe for production. No known bugs at this time. Nearing ready for merge
Custom World Gens are now supported. Conversion from pre 1.13 worlds is supported and async!

**BETA TESTERS:** 
Latest builds of this will be provided (as Paperclip patchers) here: https://aikar.co/async.jar

**DISCLAIMER:**
Again this is beta - please ensure you have backups.

These builds are now to the point they should be safe to run in production.
There has not been any reported issues yet of any data loss.

However, bugs could lead to server crashes, but they should mostly be worked out now.

**TODO LIST:**
SEE: https://github.com/PaperMC/Paper/projects/2

**DONATE:**
Lots of effort has went into making this happen. Want to say thanks and donate? 
[Foo](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=payments%40starlis%2ecom&lc=US&item_name=AsyncChunksDonation&no_note=0&currency_code=USD&bn=PP%2dDonationsBF%3abtn_donate_SM%2egif%3aNonHostedGuest)

### CODE REVIEW

**Comment by aikar** [comment]
build to jar is updated with latest updates

There is still an issue with entities created during world generation though thats going to be tough to resolve due to bad logic on mojangs part

**Comment by Lachney** [comment]
Some logs I got while testing using async.jar
https://gist.github.com/Lachney/b65f339988bce6b4468a19948d967893
https://gist.github.com/Lachney/73e3e0833a7b104e4708692f8ee9c8e7
https://gist.github.com/Lachney/930ebff98afa4d63511a5f37e27c41df

**Comment by aikar** [comment]
New beta build up!

All known bugs should now be fixed. I need testers to find bugs!
But please, test on a COPY of your production world, do not run this on your live server until it looks like no more major bugs are seen.

**Comment by aikar** [comment]
New build uploaded with even better performance for chunk conversions, as I've solved hopefully all the concurrency concerns with datafixers.
Also made spawn chunks load async.

**Comment by andris155** [comment]
Server crash:
https://pastebin.com/mAmhZqnd

**Comment by aikar** [comment]
@andris155 does it happen consistently? Thanks that's helpful

**Comment by aikar** [comment]
Note it was due to a random teleport

**Comment by andris155** [comment]
Sometimes when players use Randomteleport sign. I tested other randomteleport plugin, but same error.

**Comment by andris155** [comment]
Couldn't generate chunk (plotworld:19,-16) error (Generator: PlotSquared)
https://pastebin.com/2rzNdkk6

**Comment by aikar** [comment]
Beta Build updated with more fixes to some of the reported issues by @andris155 

The PlotSquared one is not fixed yet though, as that one is trickier and is going to require keeping those on main thread it seems.

**Comment by aikar** [comment]
At this point, if you do not use any plugin providing custom world generators, and you take adequate backups and are ok with any risks running these builds, I think we've reached the point it is testable in production on your real world.

Again, please ensure you back up correctly and understand there still could be issues.

https://aikar.co/async.jar - Updated to Paper 231

**Comment by aikar** [comment]
New Beta Build up with even more improvements based on latest paper 238 with the Tab Complete fix:
https://aikar.co/async.jar  - BETA DOWNLOAD LINK

Again this is beta - please ensure you have backups.

As said above, these builds are now to the point they should be safe to run in production.
There has not been any reported issues yet of any data loss.

However, bugs could lead to server crashes, but they should mostly be worked out now.

No known bugs outside of using Custom World Generators such as PlotSquared.
Do not use this build with PlotSquared if you have plots (re)generate

**Comment by andris155** [comment]
I donated this project for PlotSquared fix. :)

**Comment by NFRCG** [comment]
Hello, found an error saving a chunk on the latest build: https://gist.github.com/NolanMC/0d5b7ed42a25c122f0bca2512c2fca32

/version output: 10.09 09:17:24 [Server] INFO This server is running Paper version git-Paper-"143c2e0e" (MC: 1.13.1) (Implementing API version 1.13.1-R0.1-SNAPSHOT)
10.09 09:17:24 [Server] INFO Previous version: git-Paper-"2550a2a1" (MC: 1.13.1)
10.09 09:17:24 [Server] INFO Error obtaining version information

I obtained this error by simply logging into my previous location after the restart which put the latest jar in. I did not take any additional actions.
EDIT: Running /save-all doesn't replicate the error.

**Comment by tlaundal** [comment]
Tested out latest beta (downloaded 90 minutes before this comment) with 20 players.
Got hangs after 30 seconds, coming right after each other. Stopped the server with /stop.

See logs here: https://dpaste.de/fNMq
I removed some player information.

**Comment by aikar** [comment]
major performance boost in main 1.13 branch around liquids, as well as boosted priorities for main blocking chunk gens now in latest builds.

None of the other recent bugs reported fixed yet though. @totokaka  I need the full crash report. stopping it early made it lose important information :/

We need the full thread dump that prints at end. 

Download latest build linked above

**Comment by tlaundal** [comment]
Will make sure to let it run until it crashes and generates a report next time, though I think I'll wait a bit longer before I use it with all my players again.

**Comment by ghost** [comment]
Couldn't generate chunk error, didn't crash the server, but generated a 342MB log before I shut it down: https://pastebin.com/EVLE0m3K

**Comment by andris155** [comment]
```
[16:33:43] [Server thread/WARN]: Entity threw exception at world:177.5,55.0,12780.5
[16:33:43] [Server thread/WARN]: java.lang.NullPointerException
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.PathfinderNormal.b(PathfinderNormal.java:350)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.PathfinderNormal.a(PathfinderNormal.java:307)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.PathfinderNormal.a(PathfinderNormal.java:273)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.PathfinderNormal.a(PathfinderNormal.java:242)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.Navigation.a(SourceFile:191)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.Navigation.a(SourceFile:131)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.NavigationAbstract.o(NavigationAbstract.java:219)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.NavigationAbstract.d(NavigationAbstract.java:176)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.NavigationSpider.d(SourceFile:63)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.EntityInsentient.doTick(EntityInsentient.java:678)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.EntityLiving.k(EntityLiving.java:2326)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.EntityInsentient.k(EntityInsentient.java:511)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.EntityMonster.k(EntityMonster.java:22)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.EntityLiving.tick(EntityLiving.java:2174)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.EntityInsentient.tick(EntityInsentient.java:249)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.EntityMonster.tick(EntityMonster.java:26)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.EntitySpider.tick(EntitySpider.java:45)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.World.entityJoinedWorld(World.java:1562)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.World.g(World.java:1529)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.World.tickEntities(World.java:1336)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.WorldServer.tickEntities(WorldServer.java:615)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.MinecraftServer.b(MinecraftServer.java:1104)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.DedicatedServer.b(DedicatedServer.java:428)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.MinecraftServer.a(MinecraftServer.java:956)
[16:33:43] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.MinecraftServer.run(MinecraftServer.java:854)
[16:33:43] [Server thread/WARN]: 	at java.lang.Thread.run(Thread.java:748)
```

**Comment by zachbr** [comment]
@aikar

**Comment by aikar** [comment]
new build up that will fix custom world generators such as PlotSquared, as well as pulls in master for the Pathfinder NPE fix.

**Comment by aikar** [comment]
new build up that should fix that structure save issue reported, though that was looking to be extremely rare.

**Comment by NFRCG** [comment]
Latest build, the issue with structure saving still persists for me
Entire startup log: https://gist.github.com/NolanMC/836c1596982fb7ec97988a356fb30406

The error was occurring immediately after the server fully loaded, and happened without me even logging into the server. It is appearing to only be on startup, and not at any other time. 

/version output: 12.09 12:04:33 [Server] INFO This server is running Paper version git-Paper-"740f8d4a" (MC: 1.13.1) (Implementing API version 1.13.1-R0.1-SNAPSHOT)
12.09 12:04:33 [Server] INFO Previous version: git-Paper-"1b0c2734" (MC: 1.13.1)
12.09 12:04:33 [Server] INFO Checking version, please wait...
12.09 12:04:34 [Server] INFO Unknown version

If any other information is needed please let me know.

**Comment by aikar** [comment]
@NolanMC does it consistently happen?

Can you zip up your files in the world folder (just the first level, not the sub folders), and the files in

data/

and

region/r.0.0.mca
region/r.-1.0.mca
region/r.1.0.mca
region/r.0.-1.mca
region/r.0.1.mca

assuming your spawn is near 0.0?

If you can send me zip in PM on discord or IRC

**Comment by NFRCG** [comment]
@aikar Happens on every restart. I have pmed you the zip on discord.

**Comment by Bobcycle** [comment]
Citizens seems to be having a path finding problem, says Aikar. Can't read code myself, but I'm pretty sure it's something you guys can fix! https://paste.md-5.net/tehofawabe.avrasm, whenever I try to create an NPC it gives me a NPE aswell, maybe something to note.

**Comment by aikar** [comment]
new beta build up with more structure and stronghold specific concurrency fixes.

also MAJOR user experience boost.... We are now boosting priority of any chunk nearby the player and also chunks directly in front of the player.

This will give a major boost to user experience to be even faster chunk loads.

This is especially important for worldgen.... if players are exploring, we will now generate priority in front of them, hopefully avoiding them hitting the edge of the world.

This build is on top of paper 248 released tonight which contains fix for the pathfinder issue for citizens, as well as other performance improvements.

update recommended :)

https://aikar.co/async.jar

**Comment by maxerium** [comment]
https://cdn.discordapp.com/attachments/488857005558202419/489815210836885504/image0.jpg

Players are getting glitched chunks (not being rendered, nor can they move) when teleporting around.
They're forced to rejoin upon this bug coming up.

**Comment by aikar** [comment]
new build up with fixes for chunk map issue and more fine tuning of load priority

**Comment by egg82** [comment]
Looks like a mem leak somewhere around here when generating lots of chunks. Reproducible steps below.

Method:
Download latest async Paper downloaded from above
Args: https://pastebin.com/gMBAvX0P
Plugins: WorldBorder, EssentialsX, EssentialsXChat, Vault, LuckPerms
view distance 18 in both spigot.yml and server-properties

Run these commands from console:
/wb clear
/wb set world 5000 5000 0 0
/wb world fill

Watch console and profilers as your memory starts slowly filling up until you only have a few MB of memory left. You don't even need to login!

I managed to capture some profiler magic. Here's a few screenshots.
http://prntscr.com/ku6bdl
http://prntscr.com/ku6bjh
http://prntscr.com/ku6bt0

**Comment by aikar** [comment]
new build was posted tonight that pulls latest master in, which includes a fix for that treasure map crash reported. Treasure maps will no longer spam load hundreds of chunks.

**Comment by aikar** [comment]
build updated with latest master which has some important fixes.

still no new progress on custom gen issue yet though

**Comment by aikar** [comment]
good news, got word there is not actually a custom world gen issue, was a confusion in report.

custom world gen should be ok. I've tested plot squared and void generator.

appears there is a rare concurrency bug that can make generations fail though, but only 1 report so far.

**Comment by aikar** [comment]
build updated with latest memory leak fixes from master, mentioned above by egg. it was not caused by async work, its a vanilla bug, so it's fixed completely now.

I also added some debug to diagnose source of the issue i mentioned above about the rare crash.

if you see any stack traces saying "write on main", please let me know!

**Comment by aikar** [comment]
more fixes to the expiring map stuff, should hopefully be no more crashes or log lines from it i hope...
build updated

**Comment by aikar** [comment]
woo big update.
I was able to just turn off Mojangs "Let's crash the server if 2 threads try to write at same time code", and instead..... just make them thread safe.... like a logical person would have done :|

This solves the "Writing from multiple thread" crash

This then enabled me to enable parallel chunk generation. I didn't even need to do the regionized concept.
So now, you will find chunks generate even faster....

At this point, no more critical bugs are known at this point.

Please test to help find if there are any new bugs :)

**Comment by aikar** [comment]
Well, had to revert the parallel world gen. it was causing crashes that I'm pretty confident we can't fix until mojang rewrites world gen to be thread safe.

We still get to keep the change that fixes the crash though.

Build updated

**Comment by lukeeexd** [comment]
Hi Aikar,

Back in 1.12.2 one of my staff had an issue when he teleported he’d get a huge lag spike, happened loading new chunks too, he’d often get dcd and won’t be able to join again. We tried literally everything under the sun to fix it. So since switching to 1.13.1 the issue is gone. Except once I moved to async beta the issue is back.

Now you may be thinking that it's just coincidence. So this morning I ran an async build and the same issue happened, updated to newest and same issue happened. Switch back to Paper without async and it was fine, tried async again and the issue was there, moved back to no async and issue gone.

https://cdn.discordapp.com/attachments/426506011575320577/491569679962341376/2018-09-18_04.22.08.png
https://cdn.discordapp.com/attachments/426506011575320577/491569677265403904/2018-09-18_04.21.56.png

These are pics of what happens on async

**Comment by aikar** [comment]
@lukeeexd sounds like a weird client bug. Is he on integrated graphics? Millions of people don't have the issue.

The only thing that I can see is that that we may send packets about the players position before the chunk they are in, but that from what I can see is possible in vanilla though

**Comment by lukeeexd** [comment]
@aikar Nope, hes using his Ryzen GPU. We have literally tried everything back in 1.12.2 to try fix it. He even got a new GPU and same issue. If it only happens on async build then it must be something to do with that? I'm on latest Paper build atm and it's perfect for him now

**Comment by aikar** [comment]
build updated on latest paper.

I would love to hear from people - who's running it in production?
how many players?
what plugins do you use that may mess with chunk/world data?

haven't heard of any new issue (outside of the last comment, which is client bug related, we can explore fixing it after its merged if we can figure it out) lately, so we're nearing ability to merge to master!

**Comment by pop4959** [comment]
I'm not using it in production but I just want to say I appreciate the issue you opened on WorldBorder for adding async chunk generation via the fill command, cause I just pre-generated a 20K x 20K world and it took me 51 hours. With async, I imagine that could be improved significantly.

**Comment by aikar** [comment]
@pop4959 my recommendation would be to just not pregen :)

**Comment by pop4959** [comment]
I've been using it for years because it improves server performance significantly, since no further chunk generation is required for the lifetime of the map. Of course for people who don't pre-generate, async chunk generation is a tremendous help. But for people who do, I'd be curious to see how long it would take to generate a map out using it (when the API is fully ready). :)

**Comment by aikar** [comment]
Some errors were reported to me last night related to ExpiringMap.

I've updated the build to fix those issues, though I did not have it happen in my own testing, but such is the fun of multithreading, consistency is never there!

Expiring Maps should now be thread safe.

https://aikar.co/async.jar

**Comment by aikar** [comment]
build updated with paper 275 to include the chunk conversion fix, so it should now be safe to convert worlds to 1.13 with async beta too.

I actually tested the chunk conversion fix on the async build first :P

**Comment by mibby** [comment]
Thanks for all the hard work and effort aikar! Looking forward to this being merged. :)

**Comment by DefinePvP** [comment]
I am using this in production, but due to errors I have taken it out of production.
Average player count: 10-20 usually
TPS: Steady 19.7 due to large number of survival farms.
Plugins (16): AdvancedBan*, CoreProtect, DiscordSRV*, dynmap*, EnjinMinecraftPlugin*, Essentials, Greylist*, InveRestore*, LuckPerms*, PlugMan*, SimpleTicketManager*, Statscraft*, Vault*, ViaVersion*, WorldBorder*, WorldEdit

I don't believe this is ready to be merged, I am getting this error when generating a large number of new chunks in the end by flying around with an elytra.

Error: https://pastebin.com/b7N5ruL5

**Comment by Yive** [comment]
Server stats from my testing:
- Player Count: 20 - 40 in closed whitelisted testing
- Average Entity Count: 3,500+
- Average TPS: 19 (Would be 20 if it weren't for large mob grinders)
- Average Loaded Chunks: 9,000+
- Major Plugins in use: Towny, mcMMO, SilkSpawners

Issues that I have experienced so far:
- ~Thread safety issue causing a deadlock~ Fixed
- ~Tree generation getting cut off at chunk edges~ I think it was fixed
- When multiple players with flying and increased flight speed are generating chunks, the chunk generation speed gets extremely slow

Positives about async chunks:
- Will no longer make pre-generating a huge priority like it used to be (though still is recommended to do)
- Chunk loading speed has been increased a ton compared to 1.12.2
- View distance no longer needs to be set to 4
- **Basically everything about this is better than the current chunk situation in 1.12.2 and 1.13.1**

I've only had one crash after a full week of using async chunks of which has already been fixed. I'd say this is likely production ready. Looking forward to when WorldBorder supports this so that you can pre-gen a world without worrying about TPS while players are online.

**Comment by aikar** [comment]
@DefinePvP you were using an outdated build, but also missing the real source of that error.

But there have been a few fixes since then, so please retest latest.

@Yive it's still a single queue, so yeah that will happen. But it's at least not on main thread as long as people don't enter the unfinished chunks.

**Comment by DefinePvP** [comment]
@aikar I believe this is the current build, here is the error which includes everything https://pastebin.com/nvPaFVie

**Comment by aikar** [comment]
@DefinePvP 6d0bd72e is 
```>>> git show 6d0bd72e
commit 6d0bd72e64c2330167681642c8af29e5cab4563c
Merge: 266a404c b6760850
Author: Aikar <aikar@aikar.co>
Date:   Sun Sep 16 00:26:56 2018 -0400

    Merge branch 'master' into async-chunks
    
    * master:
      fix bug in last bug fix, recursion!
```

that is not the latest.

**Comment by aikar** [comment]
https://aikar.co/async.jar?ergergreger  download with this < your browser may of cached the download.

**Comment by aikar** [comment]
I've updated my server config to send "No Cache" instructions, hopefully that'll fix cache issues.

**Comment by aikar** [comment]
@PepijnGamerz because this isn't the official branch, version command won't let you know when there has been updates.

We would of had to of done some extra leg work for that, but since we are nearing merge point, not worth it now.

**Comment by hafarooki** [comment]
What if to fix the major slowdown, pending chunk generations out of render distance are dropped if there are too many pending?

**Comment by aikar** [comment]
not needed, we already handle bumping priority of anything near/in front the player. just let it queue.

**Comment by hafarooki** [comment]
So the issue is just too many chunk generations in general, not too many already queued

**Comment by jordanwilliams1** [comment]
Server keeps crashing when trying to use this:
https://pastebin.com/raw/wvHieLja

**Comment by aikar** [comment]
Ok that's a valid bug but should be easy to fix, on it.

**Comment by jordanwilliams1** [comment]
@aikar Nice :) Went to test this tonight because I was Getting hoffiric lag on 1.13. Lag spikes up to 4 seconds long. Soon as i loaded this it seemed a bit more smooth but then crashed.

**Comment by aikar** [comment]
Build updated with fix for @jordanwilliams1's issue as well as a concurrency fix for expiring maps

**Comment by jordanwilliams1** [comment]
@aikar Has a crash twice today, last updated the jar 10 hours ago. Here's my crash: 
https://pastebin.com/raw/xmvL0XLd

**Comment by aikar** [comment]
Updated with fixes for jordans issue.

**Comment by mibby** [comment]
@aikar Is the current async build upstream with latest paper changes as of https://github.com/PaperMC/Paper/commit/cce969a59df9d433c520269f528576e37409b09a?

**Comment by aikar** [comment]
@bergerkiller got reports of errors with one of your plugins: https://pastebin.com/shyhZMV8

Can you please add support for this to your plugin.

**Comment by aikar** [comment]
Oh I wonder if its because I made the ctor package-private, let me bump that up.

**Comment by aikar** [comment]
user reported its no longer complaining when i made the class public now.

**Comment by bergerkiller** [comment]
If there happen to be issues after these changes with my plugin(s) people can open an issue ticket over at https://github.com/bergerhealer/BKCommonLib

I'm close to releasing a new 1.13.1-v2 with mostly just bugfixes and compatibility patches, so any extra changes needed for supporting these changes in paper are no problem. Just need to know about it timely :)

Class / member visibility tends to be an issue with generated code, shame the error was so cryptic though. Indeed, your changes should fix that.

**Comment by aikar** [comment]
build updated with latest master as well as a new concurrency fix. update!

**Comment by WesleyVanNeck** [comment]
awsome it works great i dident found a big bug or small one

**Comment by Lachney** [comment]
awesome work on all of this, keep it up @aikar

**Comment by bergerkiller** [comment]
@aikar One suggestion for the future: considering Paper changes the chunk provider server, and BKCommonLib does so too (using a hook) passing to the underlying implementation, I have a request to add something in paper so that this hooking is no longer required.

BKCommonLib hooks into the server's CPS to override and handle [getMobsFor](https://github.com/bergerhealer/BKCommonLib/blob/master/src/main/java/com/bergerkiller/bukkit/common/internal/hooks/ChunkProviderServerHook.java#L54). If paper can provide a similarly suited event for plugins to handle this, I can disable this hooking entirely, avoiding potential compatibility issues in the future with these enhancements.

The getMobsFor event handling enables a plugin to tweak, disable or alter what mob groups spawn on the worlds. This is an early stage of spawning entities. By doing this here, a significant performance improvement is achieved, because no entities have to be created - and then destroyed - as part of the normal spawn events Bukkit offers.

Would you be interested in providing such extra functionality to plugins? I will write a separate issue ticket for this feature request if it has your initial approval, to stay on topic of this issue ticket.

(BKCommonLib extends the CPS presently installed, so it should not cause any of these changes in paper to be cancelled out)

**Comment by mibby** [comment]
@aikar Does https://aikar.co/async.jar auto-build and host the latest changes made to the async branch? The file didn't seem to change with the new master commit merged.

**Comment by hafarooki** [comment]
It does not auto update, you need to redownload it

**Comment by aikar** [comment]
It is now updated @mibby. I had internet issues last night and could not access my server. But rebuilt and uploaded now with the light queue changes

**Comment by aikar** [comment]
Thank you to everyone who has helped test this! We have now landed this in master!

Paper builds 302+ now include Async Chunks. Please report issues as a new issue and on discord

**Comment by BBoyJD10** [comment]
Sometime citizens npcs do not show

**Comment by sgdc3** [comment]
@BBoyJD10 i noticed that too but seems unrelated to the async chunk change, as it was happening before the merge.

**Comment by viet** [comment]
Just out of curiousity, does this async chunk loading work now? I saw the config on paper.yml but not sure what's the right config or if it works yet.
